[pytorch] Add env variable to enable IPC for expandable segments

pavanbalaji opened the pull request:
Summary:
PyTorch's expandable segments IPC capability was disabled in fbcode
due to job failures (see
https://github.com/pytorch/pytorch/pull/132890).  However, some use
cases like CTran require IPC functionality for multi-process GPU
communication. This change introduces
PYTORCH_CUDA_EXPANDABLE_SEGMENTS_IPC environment variable to allow
opt-in enablement of IPC handle types for expandable segments in
fbcode builds while maintaining backward compatibility.

IPC is enabled by default in non-fbcode builds and disabled by default
in fbcode builds (existing behavior). In both cases, it can be
explicitly enabled by setting PYTORCH_CUDA_EXPANDABLE_SEGMENTS_IPC=1.

Test Plan: CI

Differential Revision: D88274246

albanD reviewed (commented):
A test to ensure this works as expected would be great.

ngimel commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -419,14 +419,29 @@ struct ExpandableSegment {
+      static const bool default_enable_ipc = true;
 #endif
+      static const auto env =
+          c10::utils::get_env("PYTORCH_CUDA_EXPANDABLE_SEGMENTS_IPC");
```
Good point!

pavanbalaji committed:
[pytorch] Add env variable to enable IPC for expandable segments (#169487)

Summary:
PyTorch's expandable segments IPC capability was disabled in fbcode
due to job failures (see
https://github.com/pytorch/pytorch/pull/132890).  However, some use
cases like CTran require IPC functionality for multi-process GPU
communication. This change introduces
PYTORCH_CUDA_EXPANDABLE_SEGMENTS_IPC environment variable to allow
opt-in enablement of IPC handle types for expandable segments in
fbcode builds while maintaining backward compatibility.

IPC is enabled by default in non-fbcode builds and disabled by default
in fbcode builds (existing behavior). In both cases, it can be
explicitly enabled by setting PYTORCH_CUDA_EXPANDABLE_SEGMENTS_IPC=1.


Test Plan: CI

Reviewed By: ngimel

Differential Revision: D88274246

Pulled By: fduwjj
